forge/src/trace/mod.rs

```diff
@@ -0,0 +1,357 @@
+                let node =
+                    CallTraceNode { parent: Some(entry), trace: new_trace, ..Default::default() };
+                self.arena.push(node);
+                self.arena[entry].children.push(id);
+
+                id
+            }
+            // We haven't found the parent node, go deeper
+            _ => self.push_trace(
+                *self.arena[entry].children.last().expect("Disconnected trace"),
+                new_trace,
+            ),
+        }
+    }
+
+    pub fn addresses_iter(&self) -> impl Iterator<Item = (&Address, Option<&Vec<u8>>)> {
+        self.arena.iter().map(|node| {
+            let code = if node.trace.created {
+                if let RawOrDecodedReturnData::Raw(bytes) = &node.trace.output {
+                    Some(bytes)
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+
+            (&node.trace.address, code)
+        })
+    }
+}
+
+const PIPE: &str = "  │ ";
+const EDGE: &str = "  └─ ";
+const BRANCH: &str = "  ├─ ";
+const CALL: &str = "→ ";
+const RETURN: &str = "← ";
+
+impl fmt::Display for CallTraceArena {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fn inner(
+            arena: &CallTraceArena,
+            writer: &mut (impl Write + ?Sized),
+            idx: usize,
+            left: &str,
+            child: &str,
+        ) -> fmt::Result {
+            let node = &arena.arena[idx];
+
+            // Display trace header
+            writeln!(writer, "{}{}", left, node.trace)?;
+
+            // Display logs and subcalls
+            let left_prefix = format!("{}{}", child, BRANCH);
+            let right_prefix = format!("{}{}", child, PIPE);
+            for child in &node.ordering {
+                match child {
+                    LogCallOrder::Log(index) => {
+                        let mut log = String::new();
+                        write!(log, "{}", node.logs[*index])?;
+
+                        // Prepend our tree structure symbols to each line of the displayed log
+                        log.lines().enumerate().try_for_each(|(i, line)| {
+                            writeln!(
+                                writer,
+                                "{}{}",
+                                if i == 0 { &left_prefix } else { &right_prefix },
+                                line
+                            )
+                        })?;
+                    }
+                    LogCallOrder::Call(index) => {
+                        inner(arena, writer, node.children[*index], &left_prefix, &right_prefix)?;
+                    }
+                }
+            }
+
+            // Determine color for return arrow
+            let color = if node.trace.address == *CHEATCODE_ADDRESS {
+                Colour::Blue
+            } else if node.trace.success {
+                Colour::Green
+            } else {
+                Colour::Red
+            };
```

Comment:
you can prob move this to a helper `fn color(address: Address)` given it appears below too